### PR TITLE
fix: ds discard popup issue fixes

### LINKED ISSUE #18962: [Bug]: Even without editing the user is displayed with Pop up in Auth API
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When the user click on edit of existing Datasource it is observed that the user without making any change the pop up is displayed to the user

### Steps To Reproduce

1) Ensure there is existing datasource
2) Click on Edit option 
3) Ensure not make any change 
Observe that the pop up is displayed to the user

### Public Sample App

_No response_

### Issue video log

https://www.loom.com/share/ebe56e716afb4bc18a3e6e8a8c6c200f

### Version

Release

### PR DESCRIPTION
**Pull request by sneha122**
## Description

This PR includes following changes:
- With latest datasource autosave improvements, we do not save the datasource immediately but save it on explicit click of save button, so in case user has not saved any changes and tries to leave the page, we have added a popup to inform users that they have unsaved changes and whether they would like to save them or not.
- The issue was in case of postgres and authenticated API datasource, this popup was getting seen even when the user has not made any changes in the datasource configuration. This PR solves that issue.
- The unsaved changes popup needs to be shown only when user has made any new changes in the datasource form

TL;DR
- We have used redux form's isDirty method to check if user has made any new updates to the form or not. This isDirty compares initial value of datasource form with current form value and if current form value is different it shows the unsaved changes popup
- The issue occurred because in case of postgres and authenticated API datasource, we initialise 1 default pair of host address and port (postgres), and default pairs of headers and query parameters (Authenticated API). These initial default changes made the form dirty and so user used to see the popup even when they have not explicitly made any changes.
- This PR fixes the issue by setting initial form value with these defaults so they do not make the form dirty.

Fixes #18962 , #18998 

Media

https://user-images.githubusercontent.com/30018882/208931098-b570e3c4-10bc-4b76-bd54-531ccf869436.mov

> A video or a GIF is preferred. when using Loom, don’t embed because it looks like it’s a GIF. instead, just link to the video

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- Manual
- Cypress

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+let dsName: any;
+
+const agHelper = ObjectsRegistry.AggregateHelper,
+  dataSources = ObjectsRegistry.DataSources;
+
+const testString = "test";
+
+describe("datasource unsaved changes popup shows even without changes", function() {
+  // In case of postgres and other plugins, host address and port key values are initialized by default making form dirty
+  it("1. Create postgres datasource, save it and edit it and go back, now unsaved changes popup should not be shown", () => {
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:12]: Add Bug Id to `it` title. Also add a note for why dataSources.CreateDataSource() is not used.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      cy.wait(1000);
+      dataSources.EditDatasource();
+      agHelper.GoBack();
+      agHelper.AssertElementVisible(dataSources._activeDS);
+      dataSources.DeleteDatasouceFromActiveTab(dsName);
+    });
+  });
+
+  // In case of Auth DS, headers, query parameters and custom query parameters are being initialized, which makes form dirty
+  it("2. Create REST API datasource, save it and edit it and go back, now unsaved changes popup should not be shown", () => {
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:29]: Add Bug Id.

File: app/client/cypress/support/Pages/DataSources.ts [hunk 3]
```diff
@@ -779,4 +783,37 @@ export class DataSources {
...
+    authURL: string,
+  ) {
+    this.agHelper.GetNClick(datasourceEditor.authType);
+    this.agHelper.GetNClick(datasourceEditor.OAuth2);
+    this.agHelper.GetNClick(datasourceEditor.grantType);
+    this.agHelper.GetNClick(datasourceEditor.authorizationCode);
+    cy.get(datasourceEditor.accessTokenUrl).type(accessTokenUrl);
+    cy.get(datasourceEditor.clienID).type(clientId);
+    cy.get(datasourceEditor.clientSecret).type(clientSecret);
+    cy.get(datasourceEditor.authorizationURL).type(authURL);
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/DataSources.ts:805]: Can we please use this.agHelper.TypeText(locator, newValue) here?

File: app/client/cypress/support/Pages/DataSources.ts [hunk 4]
```diff
@@ -779,4 +783,37 @@ export class DataSources {
         0,
       );
   }
+
+  public FillAuthenticatedAPIForm() {
+    const URL = datasourceFormData["authenticatedApiUrl"];
+    cy.get(datasourceEditor.url).type(URL);
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/DataSources.ts:789]: Can we please use this.agHelper.TypeText(locator, newValue) here?

File: app/client/cypress/support/Pages/DataSources.ts [hunk 5]
```diff
@@ -779,4 +783,37 @@ export class DataSources {
...
+
+  public EditDSFromActiveTab(dsName: string) {
+    this.agHelper.GetNClick(this._editDatasourceFromActiveTab(dsName));
+  }
+
+  public FillMongoDatasourceFormWithURI(uri: string) {
+    cy.xpath(this._dropdownTitle("Use Mongo Connection String URI")).click(); //to expand the dropdown
+    cy.xpath(this._visibleTextSpan("Yes"))
+      .last()
+      .click({ force: true }); //to select the new value
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/DataSources.ts:816]: Can we check if we can use -     dataSources.ValidateNSelectDropdown("Use Mongo Connection String URI", "No", "Yes"); - can be used instead of these?

File: app/client/cypress/support/Pages/DataSources.ts [hunk 6]
```diff
@@ -779,4 +783,37 @@ export class DataSources {
...
+    cy.get(this._inputControl).type(uri);
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/DataSources.ts:817]: Instead try to use - `this.agHelper.TypeText(this.locator._inputFieldByName("Connection String URI")+ "//input", uri);`

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    dataSources.NavigateToDSCreateNew();
+    agHelper.GenerateUUID();
+    cy.get("@guid").then((uid) => {
+      dataSources.CreatePlugIn("Authenticated API");
+      dsName = "AuthDS" + uid;
+      agHelper.RenameWithInPane(dsName, false);
+      dataSources.FillAuthenticatedAPIForm();
+      dataSources.SaveDatasource();
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:37]: Can `'Authenticated API' `be also added to `CreateDataSource() `method similar to Postgres, MySql, Mongo?

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+      // Edit DS from active tab and add oauth2 details
+      dataSources.EditDSFromActiveTab(dsName);
+      dataSources.AddOAuth2AuthorizationCodeDetails(
+        testString,
+        testString,
+        testString,
+        testString,
+      );
+      dataSources.updateDatasource();
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:55]: Reame method name to PascalCase - `dataSources.UpdateDatasource`

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+      // Edit datasource, change connection string uri param and click on back button
+      dataSources.EditDatasource();
+      dataSources.FillMongoDatasourceFormWithURI(testString);
+      agHelper.GoBack();
+
+      // Assert that popup is visible
+      agHelper.AssertElementVisible(dataSources._datasourceModalDoNotSave);
+      agHelper.AssertElementVisible(dataSources._datasourceModalSave);
+      agHelper.GetNClick(dataSources._datasourceModalSave);
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:88]: Can we use - `dataSources.SaveDSFromDialog()` & add the 2 AssertElementVisible validations into that method - to make it more robust?

File: app/client/cypress/support/Pages/DataSources.ts (hunk 4 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/support/Pages/DataSources.ts:789]: Bring the datasourceEditor.url locator to this file & use that variable. Remove `import datasourceEditor` from this file

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts (hunk 8 above)
Comment (sneha122) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:55]: @Aishwarya-U-R updateDatasource method is not a new method which i have added, it has existed from before, so do we still need to update it?

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts (hunk 1 above)
Comment (sneha122) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:12]: CreateDatasource fill the form with data, I did not want to fill the form, wanted to create the plugin with default data and save it, hence used create plugin is used.

File: app/client/cypress/support/Pages/DataSources.ts (hunk 3 above)
Comment (sneha122) [reply app/client/cypress/support/Pages/DataSources.ts:805]: @Aishwarya-U-R Do we have any documentation for helper functions?

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts (hunk 7 above)
Comment (sneha122) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:37]: Yes it can be but since I did not want to fill the form and test the datasource, I just used CreatePlugIn instead

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts (hunk 1 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:12]: Here - I meant adding the note into spec file - so tomorrow someone see this spec & do not replace with `dataSources.CreateDataSource() `method without looking deeply.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts (hunk 7 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:37]: I still see `dataSources.FillAuthenticatedAPIForm();` called - which fills the form right?

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts (hunk 8 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:55]: Yes - please Rename Symbol- so the name along withs references are updated rightly.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts (hunk 7 above)
Comment (sneha122) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/BugTests/DSDiscardBugs_spec.ts:37]: yeah that function only fills in the url which is required parameter, it doesnt fill other information though

**Comment by sneha122** [comment]
@Aishwarya-U-R Resolved all code review comments. Please check

**Review by Aishwarya-U-R — COMMENTED** [review]
Cypress looks Good!

**Review by ayushpahwa — APPROVED** [review]
